server/server_test.go

```diff
@@ -2885,3 +2889,130 @@ func (m *mockResponseWriterConn) consumeQueryResultStream() ([]queryResultStream
 	}
 	return result, nil
 }
+
+func TestAuthenticationTLS(t *testing.T) {
+	ctx := context.Background()
+	store := inmem.New()
+	m, err := plugins.New([]byte{}, "test", store)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := m.Start(ctx); err != nil {
+		t.Fatal(err)
+	}
+
+	txn := storage.NewTransactionOrDie(ctx, store, storage.WriteParams)
+
+	authzPolicy := `package system.authz
+import input.identity
+default allow = false
+allow {
+	identity = "CN=my-client"
+}`
+
+	if err := store.UpsertPolicy(ctx, txn, "test", []byte(authzPolicy)); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := store.Commit(ctx, txn); err != nil {
+		t.Fatal(err)
+	}
+
+	caCertPEM, err := ioutil.ReadFile("testdata/ca.pem")
+	if err != nil {
+		t.Fatal(err)
+	}
+	pool := x509.NewCertPool()
+	if ok := pool.AppendCertsFromPEM(caCertPEM); !ok {
+		t.Fatal("failed to parse CA cert")
+	}
+	cert, err := tls.LoadX509KeyPair("testdata/server-cert.pem", "testdata/server-key.pem")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	server, err := New().
+		WithAddresses([]string{"https://127.0.0.1:8182"}).
+		WithStore(store).
+		WithManager(m).
+		WithCertificate(&cert).
+		WithCertPool(pool).
+		WithAuthentication(AuthenticationTLS).
+		WithAuthorization(AuthorizationBasic).
+		Init(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	listeners, err := server.Listeners()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	go func() {
+		// cleanup isn't our strong suit here
```

Comment:
Yikes. I feel like this illustrates a problem with the server itself. We should think about a way to fix this (not required in this PR). Could you include a comment of what's going on here? I think this means we don't expect the server to ever exit.